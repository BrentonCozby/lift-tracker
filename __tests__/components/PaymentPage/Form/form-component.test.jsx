import React from 'react'
import 'raf/polyfill'
import renderer from 'react-test-renderer'
import Enzyme, { shallow, mount } from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import configStore from '../../../../src/js/store.js'
import ConnectedForm, { Form } from '../../../../src/js/components/PaymentPage/Form/form-component.jsx'

Enzyme.configure({ adapter: new EnzymeAdapter() })

jest.mock('react-stripe-elements', () => {
    return {
        CardElement() {
            return <div className="CardElement"></div>
        },
        injectStripe(component) {
            return component
        }
    }
})

describe('Form', () => {
    let initialState
    let initialProps
    let store
    let createTokenSpy

    function initializeComponent(props) {
        const $component = shallow(<Form {...props} />)

        const componentTree = renderer.create(
            <Router>
                <Form {...props} />
            </Router>
        )

        return {
            props,
            $component,
            componentTree,
            componentJSON: componentTree.toJSON()
        }
    }

    function initializeConnectedComponent(props, state) {
        store = configStore(state)

        const $component = mount(
            <Provider store={store}>
                <Router>
                    <ConnectedForm {...props} />
                </Router>
            </Provider>
        )

        const componentTree = renderer.create(
            <Provider store={store}>
                <Router>
                    <ConnectedForm {...props} />
                </Router>
            </Provider>
        )

        return {
            props,
            $component,
            componentTree,
            componentJSON: componentTree.toJSON()
        }
    }

    beforeEach(() => {
        createTokenSpy = jest.fn(() => {
            return Promise.resolve({
                token: 'token'
            })
        })

        initialProps = {
            stripe: {
                createToken: createTokenSpy
            }
        }
    })

    test('renders payment form', () => {
        const { componentJSON } = initializeComponent(initialProps)

        expect(componentJSON).toMatchSnapshot()
    })

    test('creates a stripe token when the form is submitted', () => {
        const instance = initializeComponent(initialProps).$component.instance()

        instance.handleSubmit({
            preventDefault: () => {}
        })

        expect(createTokenSpy).toHaveBeenCalledWith({ type: 'card', name: 'Test User' })
    })

    test('renders payment form (connected)', () => {
        initialState = {}

        const { componentJSON } = initializeConnectedComponent(initialProps, initialState)

        expect(componentJSON).toMatchSnapshot()
    })
})