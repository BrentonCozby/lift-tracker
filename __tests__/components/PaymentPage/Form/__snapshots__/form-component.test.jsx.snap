// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders payment form (connected) 1`] = `
<form
  className="PaymentForm"
  onSubmit={[Function]}
>
  <div
    className="CardElement"
  />
  <input
    type="submit"
    value="submit"
  />
</form>
`;

exports[`Form renders payment form 1`] = `
<form
  className="PaymentForm"
  onSubmit={[Function]}
>
  <div
    className="CardElement"
  />
  <input
    type="submit"
    value="submit"
  />
</form>
`;
